Users/binah/access-control/access-control.cs

```diff
@@ -0,0 +1,320 @@
+    }
+
+
+    public class Visitor
+    {
+        public string Name { get; internal set; }
+        public string Handle { get; internal set; }
+        public AgentPrivate Agent { get; internal set;  }
+        public bool isAdmin { get; internal set; }
+        public bool isBanned { get; internal set; }
+    }
+
+    private Dictionary<string, Visitor> Visitors = new Dictionary<string, Visitor>();
+
+    bool IsAdmin(AgentPrivate agent)
+    {
+        if (DebugLogging) Log.Write(agent.AgentInfo.Handle + " IsAdmin: " + Admins.Contains(agent.AgentInfo.Handle.ToLower()));
+        return agent != null && Admins.Contains(agent.AgentInfo.Handle.ToLower());
+    }
+
+    bool IsBanned(AgentPrivate agent)
+    {
+        if (DebugLogging) Log.Write(agent.AgentInfo.Handle + " IsBanned: " + Banned.Contains(agent.AgentInfo.Handle.ToLower()));
+        return Banned.Contains(agent.AgentInfo.Handle.ToLower());
+    }
+
+    private void TrackUser(SessionId userId)
+    {
+        AgentPrivate agent = ScenePrivate.FindAgent(userId);
+        string handle = agent.AgentInfo.Handle.ToLower();
+        Visitor visitor;
+
+        if (DebugLogging) Log.Write(handle, " has entered.");
+
+        //if visitor is not found in list add them to the list
+        if (Visitors.TryGetValue(handle, out visitor))
+        {
+            if (DebugLogging) Log.Write("Visitor entry has been found in log.");
+        } else
+        {
+            //log user entry
+            LogEntry(agent);
+        }
+
+        if (!IsBanned(agent))
+        {
+            if(DoorsOpen || IsAdmin(agent))
+            {
+                SetAccess(agent);
+            } else
+            {
+                StartCoroutine(AskForEntry, agent);
+            }
+        } else
+        {
+            IEventSubscription timerEvent = Timer.Create(TimeSpan.FromSeconds(1), () => { Bannish(agent); });
+            Bannish(agent);
+        }
+
+    }
+
+    private void LogEntry(AgentPrivate agent)
+    {
+        Visitor visitor = new Visitor();
+        visitor.Agent = agent;
+        visitor.Name = agent.AgentInfo.Name; //used for pretty print 
+        Visitors[agent.AgentInfo.Handle.ToLower()] = visitor;
+
+        if (DebugLogging) Log.Write("Visitor entry has been logged");
+    }
+
+    private string getLogMessage()
+    {
+        string message = "There have been " + Visitors.Count + " visitors";
+        message += "\nThe doors are open: " + DoorsOpen;
+        foreach (var visitor in Visitors)
+        {
+            message += "\nName: " + visitor.Value.Name;
+            message += "\n - isAdmin: " + visitor.Value.isAdmin;
+            message += "\n - isBanned " + visitor.Value.isBanned;
+        }
+        message += "\nThe ban list: ";
+        foreach(var b in Banned)
+        {
+            message += "\n" + b;
+        }
+
+        return message;
+    }
+
+    private void ShowLog(AgentPrivate agent)
+    {
+        agent.SendChat(getLogMessage());
```

Comment:
There is a 2048 character limit to chat messages i think. might want to take care of that since ban lists are potentially limitless, also i think we will not be able to loop through all items in a persistence table

Reply:
Maybe adding like a buffer to chop and hold text going to chat or something might work ... thats a thinker, I doubt I'll solve that for Friday but will keep in mind for longer term w/ps-api